File: bet-calculator-test-jvm/src/test/java/org/strangeforest/betcalculator/test/BetCaptureCalculationTest.java
package org.strangeforest.betcalculator.test;

import java.util.*;

import org.junit.jupiter.api.*;
import org.strangeforest.betcalculator.*;
import org.strangeforest.betcalculator.bettypes.*;
import org.strangeforest.betcalculator.interrelation.*;
import org.strangeforest.betcalculator.rules.*;
import org.strangeforest.betcalculator.util.*;

import static org.assertj.core.api.Assertions.*;

class BetCaptureCalculationTest {

	@Test
	void calculateSingleBet() {
		var bet = new Bet(Single.INSTANCE, "10", List.of(
			new BetLeg("2.0", LegStatus.Companion.getOPEN(), IRDescriptor.Companion.getNO_IR(), false)
		), List.of(), BetRules.Companion.getDEFAULT());

		var result = BetCaptureCalculator.INSTANCE.calculate(bet);

		assertThat(result.getMaxReturn()).isEqualByComparingTo(new Decimal("20"));
	}

	@Test
	void calculateAccumulatorBet() {
		var bet = new Bet(Accumulator.INSTANCE, "2", List.of(
			new BetLeg("2.0", LegStatus.Companion.getOPEN(), IRDescriptor.Companion.getNO_IR(), false),
			new BetLeg("3.0", LegStatus.Companion.getOPEN(), IRDescriptor.Companion.getNO_IR(), false)
		), List.of(), BetRules.Companion.getDEFAULT());

		var result = BetCaptureCalculator.INSTANCE.calculate(bet);

		assertThat(result.getMaxReturn()).isEqualByComparingTo(new Decimal("12"));
	}

	@Test
	void calculateDoublesBet() {
		var bet = new Bet(Doubles.INSTANCE, "1", List.of(
			new BetLeg("2.0", LegStatus.Companion.getOPEN(), IRDescriptor.Companion.getNO_IR(), false),
			new BetLeg("3.0", LegStatus.Companion.getOPEN(), IRDescriptor.Companion.getNO_IR(), false),
			new BetLeg("4.0", LegStatus.Companion.getOPEN(), IRDescriptor.Companion.getNO_IR(), false)
		), List.of(), BetRules.Companion.getDEFAULT());

		var result = BetCaptureCalculator.INSTANCE.calculate(bet);

		assertThat(result.getMaxReturn()).isEqualByComparingTo(new Decimal("26"));
	}

	@Test
	void calculateBigPermsBet() {
		var bet = new Bet(new Perms(10), "1", List.of(
			new BetLeg("1.1", LegStatus.Companion.getOPEN(), IRDescriptor.Companion.getNO_IR(), false),
			new BetLeg("1.2", LegStatus.Companion.getOPEN(), IRDescriptor.Companion.getNO_IR(), false),
			new BetLeg("1.4", LegStatus.Companion.getOPEN(), IRDescriptor.Companion.getNO_IR(), false),
			new BetLeg("1.6", LegStatus.Companion.getOPEN(), IRDescriptor.Companion.getNO_IR(), false),
			new BetLeg("1.8", LegStatus.Companion.getOPEN(), IRDescriptor.Companion.getNO_IR(), false),
			new BetLeg("2.0", LegStatus.Companion.getOPEN(), IRDescriptor.Companion.getNO_IR(), false),
			new BetLeg("3.0", LegStatus.Companion.getOPEN(), IRDescriptor.Companion.getNO_IR(), false),
			new BetLeg("4.0", LegStatus.Companion.getOPEN(), IRDescriptor.Companion.getNO_IR(), false),
			new BetLeg("5.0", LegStatus.Companion.getOPEN(), IRDescriptor.Companion.getNO_IR(), false),
			new BetLeg("6.0", LegStatus.Companion.getOPEN(), IRDescriptor.Companion.getNO_IR(), false),
			new BetLeg("1.1", LegStatus.Companion.getOPEN(), IRDescriptor.Companion.getNO_IR(), false),
			new BetLeg("1.2", LegStatus.Companion.getOPEN(), IRDescriptor.Companion.getNO_IR(), false),
			new BetLeg("1.4", LegStatus.Companion.getOPEN(), IRDescriptor.Companion.getNO_IR(), false),
			new BetLeg("1.6", LegStatus.Companion.getOPEN(), IRDescriptor.Companion.getNO_IR(), false),
			new BetLeg("1.8", LegStatus.Companion.getOPEN(), IRDescriptor.Companion.getNO_IR(), false),
			new BetLeg("2.0", LegStatus.Companion.getOPEN(), IRDescriptor.Companion.getNO_IR(), false),
			new BetLeg("3.0", LegStatus.Companion.getOPEN(), IRDescriptor.Companion.getNO_IR(), false),
			new BetLeg("4.0", LegStatus.Companion.getOPEN(), IRDescriptor.Companion.getNO_IR(), false),
			new BetLeg("5.0", LegStatus.Companion.getOPEN(), IRDescriptor.Companion.getNO_IR(), false),
			new BetLeg("6.0", LegStatus.Companion.getOPEN(), IRDescriptor.Companion.getNO_IR(), false)
		), List.of(), BetRules.Companion.getDEFAULT());

		var result = BetCaptureCalculator.INSTANCE.calculate(bet);

		assertThat(result.getMaxReturn()).isEqualByComparingTo(new Decimal("1662209157.7375774976"));
	}
}

File: src/test/kotlin/org/strangeforest/betcalculator/BetCaptureCalculationTest.kt
package org.strangeforest.betcalculator

import kotlin.test.*
import assertk.*
import assertk.assertions.*

internal class BetCaptureCalculationTest {

   @Test
   fun calculateSingleBet() {
      val bet = Bet(
         "Single", "10", arrayOf(
            BetLeg("2.0")
         )
      )
      val result = calculateCapture(bet)
      assertThat(result.maxReturn).isEqualTo("20.0")
   }

   @Test
   fun calculateAccumulatorBet() {
      val bet = Bet(
         "Accumulator", "2", arrayOf(
            BetLeg("2.0"),
            BetLeg("3.0")
         )
      )
      val result = calculateCapture(bet)
      assertThat(result.maxReturn).isEqualTo("12.00")
   }

   @Test
   fun calculateDoublesBet() {
      val bet = Bet(
         "Doubles", "1", arrayOf(
            BetLeg("2.0"),
            BetLeg("3.0"),
            BetLeg("4.0")
         )
      )
      val result = calculateCapture(bet)
      assertThat(result.maxReturn).isEqualTo("26.00")
   }

   @Test
   fun calculateBigPermsBet() {
      val bet = Bet(
         "Perms10", "1", arrayOf(
            BetLeg("1.1"),
            BetLeg("1.2"),
            BetLeg("1.4"),
            BetLeg("1.6"),
            BetLeg("1.8"),
            BetLeg("2.0"),
            BetLeg("3.0"),
            BetLeg("4.0"),
            BetLeg("5.0"),
            BetLeg("6.0"),
            BetLeg("1.1"),
            BetLeg("1.2"),
            BetLeg("1.4"),
            BetLeg("1.6"),
            BetLeg("1.8"),
            BetLeg("2.0"),
            BetLeg("3.0"),
            BetLeg("4.0"),
            BetLeg("5.0"),
            BetLeg("6.0")
         )
      )
      val result = calculateCapture(bet)
      assertThat(result.maxReturn).isEqualTo("1662209157.7375774976")
   }

   @Test
   fun calculatePermsBetWithAccumulatorGroups() {
      val bet = Bet("Perms2", "1", emptyArray())
      bet.groups = arrayOf(
         BetLegGroup("Accumulator", arrayOf(
            BetLeg("2"),
            BetLeg("3")
         )),
         BetLegGroup("Accumulator", arrayOf(
            BetLeg("4")
         )),
         BetLegGroup("Accumulator", arrayOf(
            BetLeg("5")
         ))
      )

      val result = calculateCapture(bet)

      assertThat(result.maxReturn).isEqualTo("74")
   }

   @Test
   fun calculateEachWayBet() {
      val leg = BetLeg("2")
      leg.status = LegStatus("1", "0")
      leg.status?.placeOddsFactor = "0.5"
      leg.irDescriptor = IrDescriptor(123, 12, 1)
      leg.irDescriptor?.maxWinners = null
      leg.irDescriptor?.tag = "O:Score"
      val bet = Bet("Single", "10", arrayOf(leg))
      bet.rules = BetRules()
      bet.rules.eachWayType = "EACH_WAY"

      val result = calculateCapture(bet)

      assertThat(result.maxReturn).isEqualTo("35.00")
   }
}
